### ISSUE TITLE
Windows mountpoint pipeline issue

### ISSUE THREAD

**Issue by bitlogik** [Scille/parsec-cloud#1225]
Azure pipeline for Windows mountpoint test is broken.
For this minor change : https://github.com/Scille/parsec-cloud/pull/1222/files,
the parent master passed the tests ([pipeline #20200622.2](https://dev.azure.com/Scille/parsec/_build/results?buildId=1218&view=results))
The new run pipeline failed on mountpoint tests :
`hypothesis.errors.Flaky: Inconsistent data generation! Data generation behaved differently between different runs. Is your data generation depending on external state?`
https://dev.azure.com/Scille/parsec/_build/results?buildId=1239&view=logs&j=bfccef3f-45d4-504d-9a7f-a8a74922e6bf&t=1b8d0bba-f11a-562f-077c-1aaa5fdf35ee

Related to #811 and #1131 

Maybe #1223 is fixing this, or at least alleviate this, it passed this test.
The root cause may be also a non-deterministic test, like "random" files generation.

**Comment by bitlogik** [comment]
We even face some failed windows _mountpoint tests_ in the pipeline, where the tests mountpoint hang at some point, and then Azure pipeline cancels the test after 1h00 spend on the Windows. This test takes around 3mn usually.
Windows test, including the #1208 potential fix.
```
...
tests/core/mountpoint/test_folder_operations.py::test_folder_operations 
##[error]The operation was canceled.
```
Maybe there is some async event not thrown.
Not only the test fails for no reason(?), but also the test uses 60 minutes of testing pipeline without doing anything.
We are investigating on this.

**Comment by vxgmichel** [comment]
Thanks for the report!

I started investigating this too here: https://github.com/Scille/parsec-cloud/pull/1228

Feel free to use the same approach for your tests

**Comment by bitlogik** [comment]
Thanks for the input. We use your inputs to investigate :

- remove `await _wait_for_winfsp_ready(mountpoint_path) `in winfsp_runner
- remove `pytest.xfail()` in test_winfsp
- Using Winfsp (latest master)
- Use tests loop (just 8 instead 10)

Confirm the `test_folder_operations` hangs "sometimes" :
```
tests/core/mountpoint/test_file_operations.py::test_file_operations[6-8] PASSED [ 47%]
tests/core/mountpoint/test_file_operations.py::test_file_operations[7-8] PASSED [ 47%]
tests/core/mountpoint/test_file_operations.py::test_file_operations[8-8] PASSED [ 48%]
tests/core/mountpoint/test_folder_operations.py::test_folder_operations[1-8] PASSED [ 48%]
tests/core/mountpoint/test_folder_operations.py::test_folder_operations[2-8] 
##[error]The operation was canceled. [ after 55m28s, which is 59m59s of the Windows test job) ]
```

**Comment by bitlogik** [comment]
We'll add for loop and waiting time, as you did recently.

**Comment by vxgmichel** [comment]
> Confirm the test_folder_operations hangs "sometimes" 

Are you sure it actually hangs? Repeating those tests can get the run time for this task above an hour, this is what I observed during my investigation.

For the record, here's the lists of failing tests for two different runs without the relax sleep:
```
± bat logs/01.txt | grep FAILED
2020-06-24T19:18:57.4758897Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[2-10] FAILED [ 48%]
2020-06-24T19:22:30.6871846Z tests/core/mountpoint/test_winfstest.py::test_winfstest[00_CreateFile_Dispositions.t-7-10] FAILED [ 80%]
± bat logs/02.txt | grep FAILED
2020-06-24T21:42:54.8798099Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[3-10] FAILED [ 48%]
2020-06-24T21:43:22.5281762Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[5-10] FAILED [ 49%]
2020-06-24T21:45:39.0073849Z tests/core/mountpoint/test_winfstest.py::test_winfstest[00_CreateFile_Dispositions.t-2-10] FAILED [ 79%]
2020-06-24T21:46:40.9138093Z tests/core/mountpoint/test_base.py::test_mountpoint_revoke_access[write-2-10] FAILED [ 21%]
```

After adding the relax sleep of 10ms, a single test failed:
```
1& [:~/repos/parsec-cloud] [parsec] debug-mountpoint-tests+*
± bat logs/03.txt | grep FAILED
2020-06-25T10:33:42.0618167Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[1-5] FAILED [ 48%]
```

I'm launching run with a 20ms relax sleep to see if it makes a difference.

**Comment by bitlogik** [comment]
> Are you sure it actually hangs? Repeating those tests can get the run time for this task above an hour, this is what I observed during my investigation.

Was on the first for loop. Also if timeout, it would be at a random point, not at the exact same.

> After adding the relax sleep of 10ms

Yes, we are trying with 100ms.

Maybe there is an issue in winfspy FileSystem or in its configuration, so that in its start thread it ends before actually finish the task. Or there are following tasks (in Windows?) that are running later after this filesystem mounting. Theoretically, the best would be to catch when these operations are finished. Practically, a 100ms sleep time can make it work. Also this can be some tests in a waiting loop to make sure the mount point is ready.

**Comment by bitlogik** [comment]
The last function in fs.start is FspFileSystemStartDispatcher, is it async ? Or this starts some dispatcher workers to run the tasks?

**Comment by bitlogik** [comment]
We also face sometimes [this error](https://pastebin.com/M2yD7fcf) 

```
tests/core/mountpoint/test_winfstest.py::test_winfstest[10_GetSetSecurity.t-4-4] XFAIL [100%]

expected_failure = (AssertionError("DID NOT RAISED FileNotFoundError(2, 'The system cannot find the file specified')",), 'Traceback (most...pected_exc!r}")\nAssertionError: DID NOT RAISED FileNotFoundError(2, \'The system cannot find the file specified\')\n')
cmd = 'CreateFile %s\\36282552-0a70-4622-bdbe-3227eaa6c6f8\\bar GENERIC_WRITE 0 0 CREATE_NEW FILE_ATTRIBUTE_NORMAL 0'
expected = 'ERROR_PATH_NOT_FOUND'

>       assert errno == getattr(winerror, expected), f"Expected {expected}, got {message}"
E       AssertionError: Expected ERROR_PATH_NOT_FOUND, got The directory name is invalid.
E       assert 267 == 3
E         -267
E         +3

-> Expecting: ERROR_PATH_NOT_FOUND
-> Got: errno=267, message='The directory name is invalid.'
```

And [this one](https://pastebin.com/DDsStmQ6), which seems related :

```
tests/core/mountpoint/test_folder_operations.py::test_folder_operations FAILED [ 50%]

expected_failure = (AssertionError("DID NOT RAISED FileNotFoundError(2, 'No such file or directory')",), 'Traceback (most recent call las... RAISED {self.expected_exc!r}")\nAssertionError: DID NOT RAISED FileNotFoundError(2, \'No such file or directory\')\n')
```

**Comment by vxgmichel** [comment]
In the end sleeping doesn't help much, here's the results with 20ms:
```python
2020-06-25T11:32:11.8374639Z tests/core/mountpoint/test_base.py::test_mountpoint_revoke_access[read-2-5] FAILED [ 20%]
2020-06-25T11:32:13.2076477Z tests/core/mountpoint/test_base.py::test_mountpoint_revoke_access[read-4-5] FAILED [ 20%]
2020-06-25T11:43:17.2481003Z tests/core/mountpoint/test_winfstest.py::test_winfstest[00_CreateFile_Dispositions.t-2-5] FAILED [ 79%]
2020-06-25T11:43:38.1452248Z tests/core/mountpoint/test_winfstest.py::test_winfstest[08_CreateFile_Delete.t-4-5] FAILED [ 95%]
2020-06-25T11:53:07.6717936Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[2-5] FAILED [ 48%]
2020-06-25T11:55:39.5606409Z tests/core/mountpoint/test_winfstest.py::test_winfstest[05_CreateFile_Truncation.t-4-5] FAILED [ 90%]
2020-06-25T12:07:42.6118798Z tests/core/mountpoint/test_folder_operations.py::test_folder_operations[5-5] FAILED [ 50%]
```

**Comment by bitlogik** [comment]
All failures in test_winfstest.py::test_winfstest, test_folder_operations.py::test_folder_operations, and test_winfstest.py::test_winfstest might come from this issue : https://github.com/Scille/winfspy/issues/24
We can even think that test_mountpoint_revoke_access has the same root source, in the case this test is about an invalid given directory.

**Comment by vxgmichel** [comment]
Kind of fixed by #1228, I think it's good enough for the moment.

We can investigate those issues later but at least the CI should produce consistent result now :)

**Comment by vxgmichel** [comment]
Fixed (somehow): https://github.com/Scille/parsec-cloud/pull/1228

Let's discuss about Scille/winfspy#24 in #1230, thanks!
